qmetaobject/src/qmetatype.rs

```diff
@@ -144,6 +147,45 @@ fn register_metatype_qobject<T: QObject>() -> i32 {
     }
 }
 
+pub fn register_metatype_qenum<T: QEnum + QMetaType>(
+    name: *const std::os::raw::c_char,
+) -> i32 {
+    register_metatype_common::<T>(name, T::static_meta_object(), true)
+}
+
+pub fn enum_to_qvariant<T: QEnum + QMetaType>(e: &T) -> QVariant {
+    let id: i32 = T::id();
+    let raw = e.to_raw_value() as usize;
+    cpp!(unsafe [id as "int", raw as "const void*"] -> QVariant as "QVariant" {
+        return QVariant(id, raw);
+    })
+}
+
+fn qvariant_internal_ptr(var_ptr: *mut QVariant, id: i32) -> *const c_void {
+    cpp!(unsafe [var_ptr as "QVariant*", id as "int"] -> *const c_void as "const void*" {
+        return var_ptr->canConvert(id) && var_ptr->convert(id) ? var_ptr->constData() : nullptr;
+    })
+}
+
+pub fn enum_from_qvariant<T: QEnum + QMetaType>(mut variant: QVariant) -> Option<T> {
+    let id: i32 = T::id();
+    let var_ptr = &mut variant as *mut QVariant;
+    let ptr = qvariant_internal_ptr(var_ptr, id);
+    if ptr.is_null() {
+        None
+    } else {
+        let raw = unsafe{ *(ptr as *const usize) };
```

Comment:
I'm not sure if `usize` is right there.

Reply:
It should actually be the underlying type  (the one in the repr)  You could add an associated type to the QEnum trait for that.

Reply:
`trait QEnum` as conversion to and from `u32` which is the bigger supported integer type (as it is stored in a array of u32 in the `QMetaObject` data).

But here I convert `u32` to `usize` as to store it in QVariant we need to convert it to `*const c_void`. We get an error when doing so from `u32`:
```
error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
   --> src/qmetatype.rs:159:31
    |
159 |     cpp!(unsafe [id as "int", raw as "const void*"] -> QVariant as "QVariant" {
    |                               ^^^
    |
    = note: source type: `u32` (32 bits)
    = note: target type: `[u8; 8]` (64 bits)
```
This is on the to_variant side. So I take the same way out.

Reply:
Because it should be:
```
   cpp!(unsafe [id as "int", raw as "int"] -> QVariant as "QVariant" {
￼	        return QVariant(id, &raw);
```

I'm not sure why the tests are passing so there might be something wrong.

Reply:
Well, even that would not be correct because raw should not be int, but the underlying type.  Because QVariant knows the size of the enum, as it was constructed.
So either you need to always give a size of 4 byte.  Or you need to use the proper underlying type here.